pkg/apis/core/v1alpha1/cmd_types.go

```diff
@@ -93,7 +93,14 @@ type CmdSpec struct {
 	// Restarts can happen even if the command is already done.
 	//
 	// Logs of the current process after the restart are discarded.
-	RestartOn *RestartOnSpec `json:"restartOn,omitempty" protobuf:"bytes,5,opt,name=restartOn"`
+	RestartOn *TriggerSpec `json:"restartOn,omitempty" protobuf:"bytes,5,opt,name=restartOn"`
+
+	// Indicates objects that can trigger a run of this command.
+	//
+	// If there is a StartOn rule, the cmd's process will not be started until the first
+	// StartOn trigger occurs.
+	// StartOn triggers are ignored while the cmd's process is running.
+	StartOn *TriggerSpec `json:"startOn,omitempty" protobuf:"bytes,6,opt,name=startOn"`
```

Comment:
I'm a little bit nervous about representing StartOn and RestartOn with the same data model...my personal rule of thumb is to always create a new data model unless there's seem deep foundational reason why they have to be the same

For example, I could imagine modelling StartOn as having some "After" time that tells it to ignore events before a certain time, or having RestartOn having some "restart on timeout" field that wouldn't make sense for StartOn.

thoughts?

Reply:
Yeah, makes sense. This was mostly motivated by wanting to use the same Restart/TriggerManager code for both, but 1) it's not that much code to repeat (and the introduction of StartAfter actually means their logic is slightly different now) and 2) if the repetition later becomes a problem we can deal with that in a way that doesn't expose it to the API.